File: examples/dump_merkle_tree.rs
// ===========================================================================
// dump_merkle_tree.rs — Merkle 树 golden test。
//
// 运行: cargo run --example dump_merkle_tree --release > golden_merkle_rs.txt
// 覆盖: build_tree + open_path (BLAKE3-16叶/BLAKE3-32叶/SHA2-8叶)
// ===========================================================================

use whir::engines::EngineId;
use whir::hash::{Hash, BLAKE3, SHA2, ENGINES};
use zerocopy::IntoBytes;

fn layers_for_size(size: usize) -> usize {
    if size <= 1 { return 0; }
    let mut p = 1usize; let mut k = 0usize;
    while p < size { p <<= 1; k += 1; }
    k
}

fn build_tree(hash_id: EngineId, num_leaves: usize, leaves: Vec<Hash>) -> Vec<Hash> {
    let layers = layers_for_size(num_leaves);
    let leaf_layer = 1usize << layers;
    let total = (1usize << (layers + 1)) - 1;
    let mut nodes = leaves;
    nodes.resize(total, Hash::default());
    let engine = ENGINES.retrieve(hash_id).expect("engine");
    let mut prev_off = 0usize; let mut prev_len = leaf_layer;
    let mut curr_off = leaf_layer;
    for _ in 0..layers {
        let curr_len = prev_len / 2;
        let (head, tail) = nodes.split_at_mut(curr_off);
        engine.hash_many(64, &head[prev_off..prev_off + prev_len].as_bytes(), &mut tail[..curr_len]);
        prev_off = curr_off; prev_len = curr_len; curr_off += curr_len;
    }
    nodes
}

fn open_path(num_leaves: usize, witness: &[Hash], indices: &[usize]) -> Vec<Hash> {
    let layers = layers_for_size(num_leaves);
    let mut idx: Vec<usize> = indices.to_vec();
    idx.sort_unstable(); idx.dedup();
    let mut hints = Vec::new();
    let mut layer_off = 0usize; let mut layer_len = 1usize << layers;
    while layer_len > 1 {
        let mut next = Vec::with_capacity(idx.len());
        let mut k = 0usize;
        while k < idx.len() {
            let a = idx[k];
            if k + 1 < idx.len() && idx[k + 1] == (a ^ 1) {
                next.push(a >> 1); k += 2;
            } else {
                hints.push(witness[layer_off + (a ^ 1)]);
                next.push(a >> 1); k += 1;
            }
        }
        idx = next; layer_off += layer_len; layer_len /= 2;
    }
    hints
}

fn main() {
    let dump_hash = |label: &str, h: &Hash| {
        print!("  {label} ");
        for byte in h.0.iter() { print!("{byte:02x}"); }
        println!();
    };

    println!("# SECTION merkle_tree");
    let cases: &[(&str, EngineId, usize, &[usize])] = &[
        ("blake3-16-2",  BLAKE3, 16, &[3, 5, 11]),
        ("blake3-32-3",  BLAKE3, 32, &[0, 1, 2, 31]),
        ("sha2-8-2",    SHA2,    8, &[2, 6]),
    ];
    for (case_idx, &(label, hash_id, num_leaves, indices)) in cases.iter().enumerate() {
        let leaves: Vec<Hash> = (0..num_leaves)
            .map(|i| {
                let mut h = [0u8; 32];
                for (j, b) in h.iter_mut().enumerate() { *b = ((i * 31 + j) & 0xFF) as u8; }
                Hash(h)
            })
            .collect();
        let witness = build_tree(hash_id, num_leaves, leaves);
        let layers = layers_for_size(num_leaves);
        let leaf_layer = 1usize << layers;
        let total = (1usize << (layers + 1)) - 1;
        let root = witness[total - 1];
        let hints = open_path(num_leaves, &witness, indices);
        println!("CASE {case_idx} {label} num_leaves={num_leaves}");
        dump_hash("root", &root);
        println!("  leaf_layer {leaf_layer}");
        println!("  num_hints {}", hints.len());
        for (i, h) in hints.iter().enumerate() { dump_hash(&format!("hint{i}"), h); }
    }
}

File: .sync/backup-before-dev-20260525-153835/cpp/tests/dump_merkle_tree.cpp
// ===========================================================================
// dump_merkle_tree.cpp — Merkle 树 golden test。
//
// 运行: ./dump_merkle_tree > golden_merkle_cpp.txt
// 对拍: diff <(tr -d '\r' < golden_merkle_rs.txt) golden_merkle_cpp.txt
//
// 测试 Merkle 树的构建 + 打开路径:
//   1. build_tree: 自底向上, 每对相邻叶子 (64 字节) 哈希成父节点 (32 字节)
//   2. open_path:  给定叶子索引, 返回验证所需的 sibling hash 列表
//
// 3 组测试用例:
//   - BLAKE3 引擎, 16 叶子, 打开索引 [3, 5, 11]
//   - BLAKE3 引擎, 32 叶子, 打开索引 [0, 1, 2, 31]  (边界索引)
//   - SHA2   引擎,  8 叶子, 打开索引 [2, 6]
//
// 叶子生成规则 (与 Rust 完全一致):
//   leaf[i][j] = (i * 31 + j) & 0xFF
//
// 对应 Rust: examples/dump_merkle_tree.rs
// ===========================================================================

#include "whir/hash/blake3_engine.hpp"
#include "whir/hash/sha2_engine.hpp"
#include "whir/protocols/merkle_tree.hpp"
#include <cstdint>
#include <cstdio>
#include <cstdlib>
#include <vector>

int main() {
    // 打印哈希为 64 字符十六进制字符串
    auto dump_hash = [](const char* label, const whir::hash::Hash& h) {
        std::printf("  %s ", label);
        for (auto byte : h) std::printf("%02x", static_cast<unsigned>(byte));
        std::printf("\n");
    };

    namespace mt = whir::protocols::merkle_tree;
    using whir::hash::Hash;

    // 引擎实例
    whir::hash::Blake3 b3e;
    whir::hash::Sha2 s2e;

    // 引擎查找函数: 根据 EngineId 返回对应的引擎引用
    auto el = [&](whir::EngineId id) -> const whir::hash::HashEngine& {
        if (id == whir::hash::ENGINE_ID_BLAKE3) return b3e;
        if (id == whir::hash::ENGINE_ID_SHA2) return s2e;
        std::abort();
    };

    // 测试用例: (标签, 引擎ID, 叶子数, 打开的索引列表)
    struct Case { const char* label;
                  whir::EngineId hid;
                  std::size_t nl;
                  std::vector<std::size_t> idx;
                };
    const std::vector<Case> cases = {
        {"blake3-16-2", whir::hash::ENGINE_ID_BLAKE3, 16, {3, 5, 11}},
        {"blake3-32-3", whir::hash::ENGINE_ID_BLAKE3, 32, {0, 1, 2, 31}},
        {"sha2-8-2",    whir::hash::ENGINE_ID_SHA2,    8, {2, 6}},
    };

    std::printf("# SECTION merkle_tree\n");
    for (std::size_t ci = 0; ci < cases.size(); ++ci) {
        const auto& c = cases[ci];

        // 生成确定性叶子: leaf[i][j] = (i*31 + j) & 0xFF
        std::vector<Hash> leaves(c.nl);
        for (std::size_t i = 0; i < c.nl; ++i)
            for (std::size_t j = 0; j < 32; ++j)
                leaves[i][j] = (uint8_t)((i * 31 + j) & 0xFF);

        // 构建 Merkle 树
        auto cfg = mt::make_config(c.hid, c.nl);
        auto w = mt::build_tree(cfg, leaves, el);
        auto layers = mt::layers_for_size(c.nl);
        auto leaf_layer = std::size_t{1} << layers;
        auto root = mt::tree_root(w);
        auto hints = mt::open_path(cfg, w,
            std::span<const std::size_t>{c.idx.data(), c.idx.size()});

        std::printf("CASE %zu %s num_leaves=%zu\n", ci, c.label, c.nl);
        dump_hash("root", root);
        std::printf("  leaf_layer %zu\n  num_hints %zu\n", leaf_layer, hints.size());
        for (std::size_t i = 0; i < hints.size(); ++i) {
            char b[16];
            std::snprintf(b, 16, "hint%zu", i);
            dump_hash(b, hints[i]);
        }
    }
    return 0;
}
